docs/proposals/locking.md

```diff
@@ -108,3 +108,198 @@ support:
+        Err error `json:"error,omitempty"`
+}
+```
+
+
+#### `git lfs unlock <path>`
+
+The `unlock` command is responsible for releasing the lock against a particular
+file. The command takes a `<path>` argument which the LFS client will have to
+internally resolve into a UUID to unlock.
+
+The API associated with this command can also be used on the server to remove
+existing locks after a push.
+
+```go
+// An UnlockRequest is sent by the client over the API when they wish to remove
+// a lock associated with the given UUID.
+type UnlockRequest struct {
+        // UUID is the identifier of the lock that the client wishes to remove.
+        UUID string `json:"uuid"`
+}
+```
+
+```go
+// UnlockResult is the result sent back from the API when asked to remove a
+// lock.
+type UnlockResult struct {
+        // Lock is the lock corresponding to the asked-about lock in the
+        // `UnlockPayload` (see above). If no matching lock was found, this
+        // field will take the zero-value of Lock, and Err will be non-nil.
+        Lock Lock `json:"lock"`
+        // Err is an optional field which holds any error that was experienced
+        // while removing the lock.
+        Err error `json:"error,omitempty"`
+}
+```
+
+Clients can determine whether or not their lock was removed by calling the
+`Active()` method on the returned Lock, if `UnlockResult.Err` is non-nil.
+
+#### `git lfs locks (-r <remote>|-b <branch|-p <path>)|(-u uuid)`
+
+For many operations, the LFS client will need to have knowledge of existing
+locks on the server. Additionally, the client should not have to self-sort/index
+this (potentially) large set. To remove this need, both the `locks` command and
+corresponding API method take several filters.
+
+Clients should turn the flag-values that were passed during the command
+invocation into `Filter`s as described below, and batched up into the `Filters`
+field in the `LockListRequest`.
+
+```go
+// Property is a constant-type that narrows fields pertaining to the server's
+// Locks.
+type Property string
+
+const (
+        UUID   Property = "uuid"
+        Branch Property = "branch"
+        // (etc) ...
+)
+
+// LockListRequest encapsulates the request sent to the server when the client
+// would like a list of locks that match the given criteria.
+type LockListRequest struct {
+        // Filters is the set of filters to query against. If the client wishes
+        // to obtain a list of all locks, an empty array should be passed here.
+        Filters []{
+                // Prop is the property to search against.
+                Prop Property `json:"prop"`
+                // Value is the value that the property must take.
+                Value string `json:"value"`
+        } `json:"filters"`
+        // Page is the numerical page of the result set to return.
+        Page int `json:"page"`
+        // Limit is the maximum number of locks to return in a single page.
+        Limit int `json:"limit"`
+}
```

Comment:
We talked about this. Another thought we had was relying on servers to sort all locks in reverse chronological, and use cursors for paging.

REQ1: `LockListRequest{}` (first page of all locks for the repository)
RES1: `LockList{Locks: []Lock{...}, HasMore: true}`
REQ2: `LockListRequest{Cursor: "lock-uuid-1"}` (first page of all locks for the repository)
RES2: `LockList{Locks: []Lock{...}, HasMore: true}` (locks after `lock-uuid-1`)

Another thought is to replace `HasMore` with `NextCursor`, so the client doesn't have to think about what the cursor should be on requests after the initial one.